{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# DESeq2_processing_for_intersect.ipynb\n",
    "## This is intended to process DESeq2 results from <br> comparisons of non-gene genomic regions, like <br> those generated by MACS2, fStitch, or Tfit\n",
    "### The chromosome position boundaries depend on a certain <br> nomenclature from the original region bedfiles"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Import libraries"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "import pandas as pd\n",
    "import math\n",
    "import matplotlib\n",
    "import matplotlib.pyplot as plt"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Open tab delimited file, split, dump into variable 'regions'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "file1 = [\"/home/lsanford/Documents/data/nascent_ATAC_comparison/\"\n",
    "         \"Sasse/ATAC/tfea/broad/output_p1e-6/\"\n",
    "         \"B2B30ATACseq-TE_v_B2B30ATACseq-TD/temp_files/DESeq.res.txt\"]\n",
    "\n",
    "with open(file1[0]) as f:\n",
    "  regions = []\n",
    "  for line in f:\n",
    "    reg = tuple(line.strip().split(\"\\t\"))\n",
    "    regions.append(reg)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Create lists for relevant variables, depending on if DESeq2 <br> regions had an \"id\" identifier on them"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "chroms = []\n",
    "begs = []\n",
    "ends = []\n",
    "means = []\n",
    "L2FC = []\n",
    "padj = []\n",
    "\n",
    "for i in range(len(regions)):\n",
    "    if regions[0][0] == \"\\\"id\\\"\":\n",
    "        if i > 0:\n",
    "            if regions[i][8] != 'NA':\n",
    "                win = regions[i][0].strip(\"\\\"\").split(\":\")\n",
    "                bounds = win[1].split(\"-\")\n",
    "                chroms.append(win[0])\n",
    "                begs.append(bounds[0])\n",
    "                ends.append(bounds[1])        \n",
    "                means.append(float(regions[i][2]))\n",
    "                L2FC.append(float(regions[i][6]))\n",
    "                padj.append(float(regions[i][8]))\n",
    "    else:\n",
    "        if i > 0:\n",
    "            if regions[i][7] != 'NA':\n",
    "                win = regions[i][0].strip(\"\\\"\").split(\":\")\n",
    "                bounds = win[1].split(\"-\")\n",
    "                chroms.append(win[0])\n",
    "                begs.append(bounds[0])\n",
    "                ends.append(bounds[1])        \n",
    "                means.append(float(regions[i][1]))\n",
    "                L2FC.append(float(regions[i][2]))\n",
    "                padj.append(float(regions[i][7]))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Create a dataframe that puts relevant variables in bed format <br> and filters to a certain cutoff"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 288,
   "metadata": {},
   "outputs": [],
   "source": [
    "de_regions = pd.DataFrame(data={'Chromosome':chroms,'coord1':begs,\n",
    "                                'coord2':ends,'Base_mean':means,\n",
    "                                'Log2FC':L2FC, 'FDR':padj})\n",
    "de_regions = de_regions[de_regions.FDR < 0.05]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Make new filename and export data in approximately bed format"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 289,
   "metadata": {},
   "outputs": [],
   "source": [
    "indir = file1.split('/')\n",
    "outdir = '/'.join(indir[0:-1])\n",
    "outfile = ''.join([outdir,'/','processed_DESeq2_results.txt'])\n",
    "de_regions.to_csv(outfile,sep=\"\\t\",index=False,index_label=False,header=False)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
